google-cloud-spanner/src/test/java/com/google/cloud/spanner/DefaultBenchmark.java

```diff
@@ -0,0 +1,282 @@
+      results.add(service.submit(() -> runBenchmarksForQueries(server, TOTAL_READS_PER_RUN)));
+    }
+    for (int i = 0; i < PARALLEL_THREADS; i++) {
+      results.add(service.submit(() -> runBenchmarkForUpdates(server, TOTAL_WRITES_PER_RUN)));
+    }
+
+    collectResultsAndPrint(service, results, TOTAL_READS_PER_RUN + TOTAL_WRITES_PER_RUN);
+  }
+
+  /**
+   * Measures the time needed to execute a burst of read and write requests.
+   *
+   * <p>Some read requests will be long-running and will cause session leaks. Such sessions will be
+   * removed by the session maintenance background task if SessionPool Option
+   * ActionOnInactiveTransaction is set as WARN_AND_CLOSE.
+   *
+   * <p>Some write requests will be long-running. The test asserts that no sessions are removed by
+   * the session maintenance background task with SessionPool Option ActionOnInactiveTransaction set
+   * as WARN_AND_CLOSE. This is because PDML writes are expected to be long-running.
+   */
+  @Benchmark
+  public void burstUpdates(final BenchmarkState server) throws Exception {
+    final DatabaseClientImpl client = server.client;
+    SessionPool pool = client.pool;
+    assertThat(pool.totalSessions())
+        .isEqualTo(server.spanner.getOptions().getSessionPoolOptions().getMinSessions());
+
+    ListeningScheduledExecutorService service =
+        MoreExecutors.listeningDecorator(Executors.newScheduledThreadPool(PARALLEL_THREADS));
+    List<ListenableFuture<List<Duration>>> results = new ArrayList<>(PARALLEL_THREADS);
+    for (int i = 0; i < PARALLEL_THREADS; i++) {
+      results.add(service.submit(() -> runBenchmarkForUpdates(server, TOTAL_WRITES_PER_RUN)));
+    }
+
+    collectResultsAndPrint(service, results, TOTAL_WRITES_PER_RUN);
+  }
+
+  private List<java.time.Duration> runBenchmarksForQueries(
+      final BenchmarkState server, int numberOfOperations) {
+    List<Duration> results = new ArrayList<>(numberOfOperations);
+    // Execute one query to make sure everything has been warmed up.
+    executeWarmup(server);
+
+    for (int i = 0; i < numberOfOperations; i++) {
+      results.add(executeQuery(server));
+    }
+    return results;
+  }
+
+  private void executeWarmup(final BenchmarkState server) {
+    for (int i = 0; i < WARMUP_REQUEST_COUNT; i++) {
+      executeQuery(server);
+    }
+  }
+
+  private java.time.Duration executeQuery(final BenchmarkState server) {
+    Stopwatch watch = Stopwatch.createStarted();
+
+    try (ResultSet rs = server.client.singleUse().executeQuery(getRandomisedReadStatement())) {
+      while (rs.next()) {
+        int count = rs.getColumnCount();
+      }
+    }
+    return watch.elapsed();
+  }
+
+  private List<java.time.Duration> runBenchmarkForUpdates(
+      final BenchmarkState server, int numberOfOperations) {
+    List<Duration> results = new ArrayList<>(numberOfOperations);
+    // Execute one update to make sure everything has been warmed up.
+    executeUpdate(server);
```

Comment:
Shouldn't this also call `executeWarmup`, so the number of warmup queries is respected for updates as well?

Reply:
Yes, added this.